src/mono/mono/metadata/reflection-cache.h

```diff
@@ -123,6 +129,24 @@ check_object_handle (MonoMemoryManager *mem_manager, MonoClass *klass, gpointer
 		MonoWeakHashTable *hash = mem_manager->weak_refobject_hash;
 		obj_handle = MONO_HANDLE_NEW (MonoObject, (MonoObject *)mono_weak_hash_table_lookup (hash, &e));
 	}
+
+	if (!mem_manager->collectible) {
+		MonoConcGHashTable *hash = mem_manager->refobject_hash;
+		if (mono_conc_g_hash_table_lookup_extended (hash, &e, &orig_e, &orig_value))
+			if (((ReflectedEntry *)orig_e)->generation < mono_metadata_update_get_thread_generation()) {
+				mono_conc_g_hash_table_remove (hash, &e);
+				free_reflected_entry ((ReflectedEntry *)orig_e);
+				obj_handle = MONO_HANDLE_NEW (MonoObject, NULL);
+			} else {
+				obj_handle = MONO_HANDLE_NEW (MonoObject, (MonoObject *)orig_value);
+			}
+		else {
+			obj_handle = MONO_HANDLE_NEW (MonoObject, NULL);
+		}
+	} else {
+		MonoWeakHashTable *hash = mem_manager->weak_refobject_hash;
+		obj_handle = MONO_HANDLE_NEW (MonoObject, (MonoObject *)mono_weak_hash_table_lookup (hash, &e));
```

Comment:
Please make an issue to add `mono_weak_hash_table_lookup_extended` and to use it over here so that reflection cache for collectible ALCs is also keeping track of generations

Reply:
Created https://github.com/dotnet/runtime/issues/85982